sycl/source/detail/event_impl.cpp

```diff
@@ -113,10 +113,26 @@ void event_impl::setComplete() {
   assert(false && "setComplete is not supported for non-host event");
 }
 
-static uint64_t inline getTimestamp() {
-  auto Timestamp = std::chrono::high_resolution_clock::now().time_since_epoch();
-  return std::chrono::duration_cast<std::chrono::nanoseconds>(Timestamp)
-      .count();
+static uint64_t inline getTimestamp(
+    const std::weak_ptr<queue_impl> &QueueWeakPtr) {
+  if (QueueImplPtr Queue = QueueWeakPtr.lock()) {
+    try {
+      return Queue->getDeviceImpl().getCurrentDeviceTime();
```

Comment:
If all we need is the device, could it be safer to save a pointer to the device instead? Since devices don't change over the life of an application, we should not have to worry about it dying either, so we can avoid `weak_ptr` and locking.

Reply:
Sounds reasonable, I have changed it to store a pointer to the device.

Reply:
I agree. We should avoid `set*()` functions that are meant for single-call initialization of members as much as possible.